locked-asset/lkmex-transfer/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+#![no_std]
+#![allow(clippy::vec_init_then_push)]
+#![feature(generic_associated_types)]
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+pub mod constants;
+use crate::constants::*;
+
+type Epoch = u64;
+
+#[derive(
+    TopEncode, TopDecode, NestedEncode, NestedDecode, Clone, ManagedVecItem, TypeAbi, Debug,
+)]
+pub struct LockedFunds<M: ManagedTypeApi> {
+    funds: ManagedVec<M, EsdtTokenPayment<M>>,
+    time: Epoch,
+}
+
+#[elrond_wasm::contract]
+pub trait LkmexTransfer {
+    #[init]
+    fn init(
+        &self,
+        token: TokenIdentifier,
+        unlock_transfer_time: Epoch,
+        epochs_cooldown_duration: Epoch,
+    ) {
+        self.unlock_transfer_time().set(unlock_transfer_time);
+        self.epochs_cooldown_duration()
+            .set(epochs_cooldown_duration);
+        self.locked_token().set(token);
+    }
+
+    #[endpoint(withdraw)]
+    fn withdraw(&self) {
+        let caller = self.blockchain().get_caller();
+        let funds = self.get_unlocked_funds(&caller);
+        self.send().direct_multi(&caller, &funds);
+        self.locked_funds(&caller).clear();
+    }
+
+    fn get_unlocked_funds(&self, address: &ManagedAddress) -> ManagedVec<EsdtTokenPayment> {
+        require!(
+            !self.locked_funds(address).is_empty(),
+            CALLER_NOTHING_TO_CLAIM
+        );
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let unlock_transfer_time = self.unlock_transfer_time().get();
+        let locked_funds = self.locked_funds(address).get();
+        require!(
+            current_epoch - locked_funds.time > unlock_transfer_time,
+            TOKENS_STILL_LOCKED
+        );
+        locked_funds.funds
+    }
+
+    #[payable("*")]
+    #[endpoint(lockFunds)]
+    fn lock_funds(&self, address: ManagedAddress) {
+        let caller = self.blockchain().get_caller();
+        self.check_address_on_cooldown(&caller);
+
+        let payments = self.call_value().all_esdt_transfers();
+        for payment in payments.iter() {
+            require!(
+                payment.token_identifier == self.locked_token().get(),
+                BAD_LOCKING_TOKEN
+            )
+        }
```

Comment:
Rewrite so the locked token ID is only read once from storage:
```suggestion
        let locked_token_id = self.locked_token().get();
        for payment in payments.iter() {
            require!(
                payment.token_identifier == locked_token_id,
                BAD_LOCKING_TOKEN
            )
        }
```